xapian-letor/api/letor_internal.cc

```diff
@@ -123,16 +113,18 @@ Letor::Internal::load_list_fvecs(const char *filename) { //directly use train.tx
         train_file >> qid;
 
         qid = qid.substr(qid.find(":")+1,qid.length());
+        train_file.ignore();
 
-        std::map<int,double> fvals;//read features
-        for(int i = 1; i < 20; ++i){
-            train_file.ignore();
+        std::vector<double> fvals;
+
+        while (train_file.peek() != 35) { // read till '#docid' is found (35 is ascii for '#')
             int feature_index;
             double feature_value;
             train_file >> feature_index;
             train_file.ignore();
             train_file >> feature_value;
-            fvals[feature_index] = feature_value;
+            fvals.push_back(feature_value);
+            train_file.ignore();
```

Comment:
I think this parsing needs a little more commenting. You're reading a double, skipping a character without validation (space?), reading an int, skipping another character…?